### PR TITLE
Power handlers

### PR DESCRIPTION
**Pull request by ericnormand**
This one is complicated to explain, but explain it I will.

Every time we needed to add a Rabbit handler, we had to edit 4 (four!!!) files and get it all right. Services that connected to many queues had gigantic queue.clj files that were a nightmare of boilerplate. I literally saw a plate boiling in one of them.

Anyway, I started to address this problem as a namespace inside of this service. I want to eventually extract it out into a library (called slingshot), but while I was iterating on it I wanted it in the same codebase. So this review is just to make sure it all makes sense. **This PR will not be merged as is.** When this PR is LGTMed, at that point I'll extract the power namespace into the slingshot library and make a new PR in this service using that library. Should be easy.

So, good reviewer, this is what I need you to tell me:

1. Is this worthy?
2. Is it general enough to work for all services?
3. Does it work? (I'm testing it on Krakenstein myself).
4. What is the average airspeed velocity of an unladen swallow?

Just a note: As I write this, there are 445 additions and 360 deletions (diff stats). However, 253 of those are just from power.clj and 53 are in power_test.clj. So 306 of those additions will be in another library. That leaves 138 additions and 360 deletions. Yowza! Imagine the savings across all the services!

I just went ahead and did all of these:

https://www.pivotaltracker.com/story/show/112614583
https://www.pivotaltracker.com/story/show/112613531
https://www.pivotaltracker.com/story/show/112612121

### CODE REVIEW

**Comment by cap10morgan** [comment]
This all seems great but I still don't understand why it doesn't just belong in kehaar? Would slingshot be usable / useful apart from kehaar? If not, what value does it provide being separate from kehaar (vs. a new namespace in kehaar that can be used when appropriate and ignored when not)?

**Comment by tie-rack** [comment]
I'm still digesting this, but I have a few concerns:

1. Side-effects during a `def` involving the network (in `def-service-handler` and friends) make me nervous. This isn't the _value_ of the var being defined by the result of a network operation, but rather a pretty straightforward def that happens to also do operations over the network unrelated to the value.
2. I feel that part of the motivation behind Kehaar was to separate RabbitMQ from the workings of the business logic namespaces. The various handlers namespaces were freed from Rabbit, but now all those namespaces have `power` pulled in and, thus, `kehaar` and `langohr`.
3. This also feels like a regression in terms of configuring queues. The queue names for a single project have gone from being in one file to being spread across multiple files.
4. It seems that `connect-rabbit` requires that all namespaces that set up rabbit channels be loaded first. That also makes me nervous.

**Comment by ericnormand** [comment]
@cap10morgan: My understanding is that Kehaar's job is really focused. It's just a core.async to rabbit bridge. These macros include more than that. So the value is maintaining that focus.

**Comment by ericnormand** [comment]
@tie-rack This is good feedback and I'm going to rework things a little. To respond to your individual points:

1. I understand your concerns with the possible side-effect in a def. I think I can move it out of the def. The side-effect only happens when the connection has already been established, which is useful for dynamic situations like running Krakenstein. But let me work on it.

2+3. This I'm not so sure about. We could still separate them out if we choose (for instance, putting all the def-service-handler calls in a single namespace) and having each one just call functions in other namespaces to do the work. It's an option we could explore. The queue names were in two files before and tracing which handler was being used on which channel was getting difficult. In our actual usage, putting the handler in one namespace and the queue setup in another was adding nothing. They were tied together at a distance.

4. Not necessarily. The macros allow handlers to be defined after the connection is established.

This new setup is not perfect yet, so thanks for the advice. But neither was the setup we have, and this solves many problems with that. For instance, the use of globally def'ed channels that could be closed but never reopened and the duplication of queue names in two places with no means of checking if they were equal. This setup also opens the door for higher-order possibilities, like a service being able to respond to a query for what queues it is listening on, the docstrings for the services, and even the schemas.

**Comment by cap10morgan** [comment]
"Just a core.async to rabbit bridge" that results in tons of boilerplate code that necessitates additional library code to undo says to me that you need additional code (with good organization, decoupling, etc.) in your "just a core.async to rabbit bridge" library, not a new library layered on top of that library. If that new library had value completely apart from kehaar (i.e. there were reasons to pull the new library into a non-kehaar project), _then_ I would support making it separate. Needing to juggle yet another library here is overhead we should carefully consider before adopting in the name of keeping kehaar focused (but prone to boilerplate).

But I'd love to hear others' thoughts on this front.

**Comment by ericnormand** [comment]
Just for the record, I don't mind putting it in Kehaar and I don't care enough to fight about it.

I've reworked it so that it does not try to connect to rabbit if you're already connected. I also put the initializer function into the metadata for the vars. That means that they are (almost) just defs with metadata. The metadata is used by the routines to connect up the queues and handlers. This is similar to the approach taken by clojure.test.

If we're just storing stuff in the metadata, I might as well store the data and not just an initializer function. That will be next.

**Comment by ericnormand** [comment]
BTW, it does indeed work in krakenstein.

**Comment by ericnormand** [comment]
Works in Krakenstein on current commit.

**Comment by ericnormand** [comment]
If there's no more comments on this, I'm going to move power.clj into kehaar.

**Comment by tie-rack** [comment]
My lack of further commentary is a function of time to comment. However, yeah, generally, moving power to a kehaar PR seems fine.

Though, `handle-errors` still doesn't belong there.

File: src/electorate_works/core.clj [hunk 1]
```diff
@@ -1,19 +1,15 @@
 (ns electorate-works.core
-    (:require [electorate-works.channels :as channels]
-              [electorate-works.queue :as queue]
-              [turbovote.resource-config :refer [config]]
+    (:require [turbovote.resource-config :refer [config]]
               [datomic-toolbox.core :as db]
               [clojure.tools.logging :as log]
-              [immutant.util :as immutant]))
+              [immutant.util :as immutant]
+              ;; we must require all handlers to let them initialize
```
Comment (tie-rack) [inline src/electorate_works/core.clj]: The state of the Rabbit system is dependent on file load order. Is this too much magic? Perhaps there's a way to attain this kind of slimming of the Rabbit setup code but still keeping what happens during initialization explicit (and thus obviously repeatable and amenable to visual inspection).

**Pull request closed without merging**